### PR TITLE
feat(cli, metro-config): environment variable support

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

> Expo supports the [standard .env resolution](https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use).

- It's nice to be able to use uncommitted values in your app, based on the environment. This feels very familiar to web developers. 
- Values that are prefixed with `EXPO_PUBLIC_` will be inlined in the bundle when bundling normally (e.g. not for Node.js).
- `.env` files are loaded into memory and applied to the process during a run. This also means that they're available in `app.config.js`.
- During development-only, environment variables are exposed on the `process.env` object (non-enumerable) to ensure they're available between fast refresh updates.

# How

- Create new package `@expo/env` which is used to hydrate env variables in a unified way. I plan to open another PR in `eas-cli` which uses this package to fill in environment variables before uploading. NOTE: environment variables that are defined in eas.json are not available in Expo CLI when building locally, but are available in the cloud since they'll be on the process, this means they effectively emulate `.env.production`.
- Update templates to gitignore local env files.
- Add basic documentation to the versioned metro guide (more to come).

# Test Plan

- [ ] E2E rendering test
- [ ] E2E Node.js rendering test
- [x] Unit tests for serializer
- [x] Tests for env package

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-dev-client/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 ### 💡 Others
 
 - Update e2e tests to use custom entry file system. ([#21643](https://github.com/expo/expo/pull/21643) by [@EvanBacon](https://github.com/EvanBacon))
+- Update export script to use Expo CLI's `export:embed` command to ensure NODE_ENV is set. ([#21983](https://github.com/expo/expo/pull/21983) by [@EvanBacon](https://github.com/EvanBacon))
```
Comment (EvanBacon) [inline packages/expo-dev-client/CHANGELOG.md:14]:
```suggestion
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 2]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
 
 See more information about **metro.config.js** in the [customizing Metro guide](/guides/customizing-metro/).
 
+## Envrionment Variables
+
+> Since SDK 49
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:9]:
```suggestion
> For SDK 49 and above
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 3]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
...
+
+Environment variables can be loaded using `.env` files. Files are loaded according to the [standard `.env` file resolution](https://github.com/bkeepers/dotenv/blob/c6e583a/README.md#what-other-env-files-can-i-use).
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:11]:
```suggestion
Environment variables can be loaded using **.env** files. These files are loaded according to the [standard **.env** file resolution](https://github.com/bkeepers/dotenv/blob/c6e583a/README.md#what-other-env-files-can-i-use).
```

[Filenames as bold text as per the writing styelguide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#filenames-as-bold-text)

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 4]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
...
+
+If you are migrating an older project to SDK +49, then you should ignore local env files by adding the following to your `.gitignore`:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:13]:
```suggestion
If you are migrating an older project to SDK 49 or above, then you should ignore local env files by adding the following to your **.gitignore**:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 5]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
 
 See more information about **metro.config.js** in the [customizing Metro guide](/guides/customizing-metro/).
 
+## Envrionment Variables
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:7]:
```suggestion
## Envrionment variables
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 6]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
...
+Environment variables can be loaded using `.env` files. Files are loaded according to the [standard `.env` file resolution](https://github.com/bkeepers/dotenv/blob/c6e583a/README.md#what-other-env-files-can-i-use).
+
+If you are migrating an older project to SDK +49, then you should ignore local env files by adding the following to your `.gitignore`:
+
+```sh .gitignore
+# local env files
+.env*.local
+```
+
+### Client Envrionment Variables
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:20]:
```suggestion
### Client envrionment variables
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 7]
```diff
@@ -4,15 +4,40 @@ title: metro.config.js
...
+
+Envrionment variables prefixed with `EXPO_PUBLIC_` will be exposed to the app at build-time. For example, `EXPO_PUBLIC_API_KEY` will be available as `process.env.EXPO_PUBLIC_API_KEY`.
+
+Envrionmnet variables will not be inlined in Node Modules.
+
+For security purposes, client environment variables are inlined in the bundle, this means that `process.env` is not an iterable object and you cannot dynamically access environment variables. Every variable must be explicitly defined, e.g. `process.env.EXPO_PUBLIC_KEY` is valid and `process.env['EXPO_PUBLIC_KEY']` is not.
+
+- Client environment variables should not contain secrets as they will be viewable in plain-text format in the production binary.
+- Use client environment variables for partially protected values like public API keys that you don't want to commit to GitHub.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:29]:
```suggestion
- Use client environment variables for partially protected values, such as public API keys you don't want to commit to GitHub.
```

**Review by amandeepmittal — COMMENTED** [review]
Left some suggestions.

**Pull request merged**